skbio/sequence/_grammared_sequence.py

```diff
@@ -749,6 +761,57 @@ def to_regex(self, within_capture=False):
 
         return re.compile(regex_string)
 
+    def to_definites(self, degenerate="wild", char=None, canonical=True):
```

Comment:
@mataton This is impressive effort! Thanks! I wanted to make a comment on the algorithm. The current code iterates over all degenerate characters, and runs `str.replace` for each of them. There may be room to optimize this. For example, you can use regular expression:

```python
import re
pattern = "[{}]".format("".join(map(re.escape, degenerates)))
sequence = pattern.sub(char, sequence)
```

Or you can use list comprehension:

```python
sequence = "".join(char if x in degenerates else x for x in sequence)
```

I don't know which one is faster. I guess regex should win. But one should do some tests to make sure.

***

Also, you wrote the code block for replacement multiple times. This can be saved with:

```python
if degenerate == "wild":
    sub_char = self.wildcard_char
elif degenerate == "gap":
    sub_char = self.default_gap_char
elif ...
for degenerate_char in degenerates:
    sequence = sequence.replace(
        degenerate_char, sub_char
    )
```

Reply:
@qiyunzhu Thank you for the feedback! Surprisingly it actually looks like the list comprehension is faster. Because it is faster and also easier to read, I think we should go with the list comprehension option. 

<img width="409" alt="Screenshot 2024-02-23 170047" src="https://github.com/scikit-bio/scikit-bio/assets/62309966/63e7a55d-6c4a-4ab2-ab6a-d23abc1c2462">

Also, I found an error with your regex solution. It needed `re.compile` at the beginning of it to avoid trying to run `sub()` on a `string` object. The version I implemented looked like this:

```python
import re
pattern = re.compile("[{}]".format("".join(map(re.escape, degenerates))))
sequence = pattern.sub(char, sequence)
```

Reply:
Good catch! And thanks for benchmarking! I agree with you that list comprehension should be the solution. Actually, I think there are faster methods. For example, once can pre-generate the regex pattern and store it as a class property.  For another example, one can directly operate on `sequence._bytes` using some vectorized operation like `np.where(np.isin(seq._bytes, seq.definites), seq._bytes, sub_char)`. This could be faster. However, I am not sure if this step can be a bottleneck. List comprehension is the concise and easy-to-understand solution.